toolkit/tools/pkg/imagecustomizerlib/liveosisobuilder.go

```diff
@@ -864,3 +894,419 @@ func createLiveOSIsoImage(buildDir, baseConfigPath string, isoConfig *imagecusto
+//     all of its contents.
+//   - The amount of disk space a file occupies depends on the block size of the
+//     host file system. If many files are smaller than a block size, there will
+//     be a lot of waste. If files are very large, there will be very little
+//     waste. It is hard to predict how much disk space a set of a files will
+//     occupy without enumerating the sizes of all the files and knowing the
+//     target block size. In this function, we use an optimistic approach which
+//     calculates the required disk space by multiplying the total file size by
+//     a safety factor - i.e. safe that it will be able t hold all the contents.
+//
+// inputs:
+//
+//   - 'rootDir':
+//     root folder to calculate its size.
+//   - 'safetyFactor':
+//     a multiplier used with the total number of bytes calculated.
+//
+// outputs:
+//
+//   - returns the size in mega bytes.
+func getDiskSizeEstimateInMBs(rootDir string, safetyFactor uint64) (size uint64, err error) {
+
+	sizeInBytes, err := getSizeOnDiskInBytes(rootDir)
+	if err != nil {
+		return 0, fmt.Errorf("failed to get folder size on disk while estimating total disk size:\n%w", err)
+	}
+
+	sizeInMBs := sizeInBytes/1024/1024 + 1
+	return sizeInMBs * safetyFactor, nil
+}
+
+// createWriteableImageFromSquashfs
+//
+//   - given a squashfs image file, it creates a writeable image with two
+//     partitions, and copies the contents of the squashfs unto that writeable
+//     image.
+//   - the squashfs image file must be extracted from a previously created
+//     LiveOS iso and is specified by the LiveOSIsoBuilder.artifacts.squashfsImagePath.
+//
+// inputs:
+//
+//   - 'buildDir':
+//     path build directory (can be shared with other tools).
+//   - 'rawImageFile':
+//     the name of the raw image to create and populate with the contents of
+//     the squashfs.
+//
+// outputs:
+//
+//   - creates the specified writeable image.
+func (b *LiveOSIsoBuilder) createWriteableImageFromSquashfs(buildDir, rawImageFile string) error {
+
+	logger.Log.Infof("Creating writeable image from squashfs (%s)", b.artifacts.squashfsImagePath)
+
+	// mount squash fs
+	squashMountDir, err := ioutil.TempDir(buildDir, "tmp-squashfs-mount-")
+	if err != nil {
+		return fmt.Errorf("failed to create temporary mount folder for squashfs:\n%w", err)
+	}
+	defer os.RemoveAll(squashMountDir)
+
+	squashMountParams := []string{b.artifacts.squashfsImagePath, squashMountDir}
+	err = shell.ExecuteLive(false, "mount", squashMountParams...)
+	if err != nil {
+		return fmt.Errorf("failed to mount squashfs:\n%w", err)
+	}
+	defer func() {
+		unmountParams := []string{squashMountDir}
+		cleanupErr := shell.ExecuteLive(false, "umount", unmountParams...)
+		if cleanupErr != nil {
+			err = fmt.Errorf("%w:\nfailed to clean-up (%s): %w", err, squashMountDir, cleanupErr)
+		}
+	}()
+
+	// estimate the new disk size
+	safeDiskSizeMB, err := getDiskSizeEstimateInMBs(squashMountDir, 2 /*safety factor*/)
```

Comment:
Why is the safety factor so large? Being off by an order of magnitude suggests that something is wrong.

Reply:
I've changed it to 1.5. It's hard to estimate the size without some more calculations. I'm trying to err on the safe side.

Reply:
btw:
I've found out that du returns the total size of disk blocks used. It does not just add the exact size of each file. So, du will return 4k for 2 byte file, for example. This makes me think we can lower the safety factor, since choosing a block size that's not 4k is not common. Open to suggestions here...